test/extensions/filters/http/common/fuzz/filter_fuzz_test.cc

```diff
@@ -0,0 +1,159 @@
+#include <chrono>
+#include <memory>
+
+#include "envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.pb.h"
+
+#include "common/config/utility.h"
+#include "common/config/version_converter.h"
+#include "common/protobuf/utility.h"
+
+#include "extensions/filters/http/well_known_names.h"
+
+#include "test/config/utility.h"
+#include "test/extensions/filters/http/common/fuzz/filter_fuzz.pb.h"
+#include "test/fuzz/fuzz_runner.h"
+#include "test/fuzz/utility.h"
+#include "test/mocks/buffer/mocks.h"
+#include "test/mocks/http/mocks.h"
+#include "test/mocks/server/mocks.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+
+class UberFilterFuzzer {
+public:
+  UberFilterFuzzer() {
+    // Need to set for both a decoder filter and an encoder/decoder filter.
+    ON_CALL(filter_callback_, addStreamDecoderFilter(_))
+        .WillByDefault(
+            Invoke([&](std::shared_ptr<Envoy::Http::StreamDecoderFilter> filter) -> void {
+              filter_ = filter;
+              filter_->setDecoderFilterCallbacks(callbacks_);
+            }));
+    ON_CALL(filter_callback_, addStreamFilter(_))
+        .WillByDefault(
+            Invoke([&](std::shared_ptr<Envoy::Http::StreamDecoderFilter> filter) -> void {
+              filter_ = filter;
+              filter_->setDecoderFilterCallbacks(callbacks_);
+            }));
+    // Ext-authz setup
+    addr_ = std::make_shared<Network::Address::Ipv4Instance>("1.2.3.4", 1111);
+    ON_CALL(connection_, remoteAddress()).WillByDefault(testing::ReturnRef(addr_));
+    ON_CALL(connection_, localAddress()).WillByDefault(testing::ReturnRef(addr_));
+    ON_CALL(callbacks_, connection()).WillByDefault(testing::Return(&connection_));
+    callbacks_.stream_info_.protocol_ = Envoy::Http::Protocol::Http2;
+  }
+
+  // This executes the methods to be fuzzed.
+  void decode(Http::StreamDecoderFilter* filter, const test::fuzz::HttpData& data) {
+    bool end_stream = false;
+
+    Http::TestHeaderMapImpl headers = Fuzz::fromHeaders(data.headers());
+    if (headers.Path() == nullptr) {
+      headers.setPath("/foo");
+    }
+    if (headers.Method() == nullptr) {
+      headers.setMethod("GET");
+    }
+    if (headers.Host() == nullptr) {
+      headers.setHost("authority");
+    }
+
+    if (data.data().size() == 0 && !data.has_trailers()) {
+      end_stream = true;
+    }
+    ENVOY_LOG_MISC(debug, "Decoding headers: {} ", data.headers().DebugString());
+    const auto& headersStatus = filter->decodeHeaders(headers, end_stream);
+    if (headersStatus != Http::FilterHeadersStatus::Continue ||
```

Comment:
I think the state machines are independent for headers and data; you can deliver data even though headers has return `StopIteration`.

Reply:
Hm, I might have interpreted this comments in https://github.com/envoyproxy/envoy/blob/master/include/envoy/http/filter.h wrong. Let me look for code examples. If it's not `Continue` or `StopIteration`, then it's `ContinueAndEndStream` ("ignore any subsequent data or trailers"), `StopAllIIterationAndBuffer` ("do not iterate for headers as well as data and trailers for current filter"), or `StopAllIterationAndWatermark` ("do not iterate for headers as well as data and trailers for the current filter").

Reply:
Oh good point, yeah, I think I thought this was only returning if not continue, seems good.